aeneas/src/ssa/SsaDebugger.v3

```diff
@@ -235,38 +426,150 @@ class DebuggerParser {
+		while (ch != '\n') {
+			sb.putc(ch);
+			ch = byte.!(System.fileRead(STDIN));
+		}
+		var input = sb.extract();
+		if (input[0] == 'y' || input[0] == 'Y') return true;
 		else return false;
 	}
 }
 class Breakpoint {
-	def render(out: StringBuilder) -> StringBuilder;
-	def disable();
-	def enable();
-}
-class LineBreakpoint(fileName: string, line: int, var on: bool) extends Breakpoint {
-	def render(out: StringBuilder) -> StringBuilder {
-		return out.put3("[%s @ %d] (%z)", fileName, line, on);
+	def ssa: SsaCheckpoint;
+	new(ssa) {
+		ssa.breakpoint = true;
 	}
-	def disable() { on = false; }
-	def enable() { on = true; }
-}
-class SsaCPBreakpoint(ssa: SsaCheckpoint) extends Breakpoint {
 	def render(out: StringBuilder) -> StringBuilder { 
 		return out.put2("%q (%z)", ssa.source.render, ssa.breakpoint);
 	}
 	def disable() { ssa.breakpoint = false; }
 	def enable() { ssa.breakpoint = true; }
 }
-class InvalidBreakpoint extends Breakpoint {
-	def render(out: StringBuilder) -> StringBuilder { return out; }
-	def disable() {}
-	def enable() {}
-}
+
+def storeLimit = 50;
+def storeRemove = 10;
+class SrcPrinter(vstFiles: Array<VstFile>, interp: SsaInterpreter) {
+	def padNum = 6;
+	def tmpBuf = TerminalBuffer.new();
+	def outBuf = TerminalBuffer.new();
+	def ends = Vector<int>.new();
+	var on: bool;
+	var traceCallee: bool;
+	var indent: int;
+
+	def set(on: bool) {
+		this.on = on;
+		if (on) reset();
+	}
+	def reset() {
+		ends.resize(0);
+		tmpBuf.reset();
+		outBuf.reset();
+	}
+	def putData(instr: SsaInstr, targetSp: int) {
+		if (!traceCallee && interp.sp != targetSp) return;
+		if (SsaCheckpoint.?(instr)) {
+			var source = SsaCheckpoint.!(instr).source;
+			var point = FilePoint.new(null, source.line, source.column);
+			extractTmpConstraint(0);
+			addEnds(outBuf.length);
+			tmpBuf.putd(source.line).tab().rjustify(' ', padNum, padNum);
+			putNewIndent(targetSp);
+			outBuf.puts(tmpBuf.extract());
+			outBuf.puts(point.extractLine(vstFiles[0].input, vstFiles[0].lineEnds)).ln();
+		} else if (SsaApplyOp.?(instr)) {
+			putVal(instr);
+		} else if (SsaReturn.?(instr)) {
+			var i = interp.stack[interp.sp - 1].ip; // Caller instr
+			if (!SsaApplyOp.?(i)) return;
+			putVal(i);
+		}
+	}
+	def addEnds(i: int) {
+		ends.put(i);
+		if (ends.length > storeLimit) {
+			var newBuf = Arrays.range(outBuf.buf, ends[storeRemove], outBuf.length);
+			outBuf.reset();
+			outBuf.puts(newBuf);
+			var newEnds = Arrays.range(ends.array, storeRemove + 1, ends.length);
+			for (j < newEnds.length) newEnds[j] -= ends[storeRemove];
+			Vectors.overwrite(ends, newEnds);
+		}
+	}
+	def putVal(instr: SsaInstr) {
+		var vtype = instr.getType();
+		if (VoidType.?(vtype) || instr.valueNum < 0) return;
+		var source = SsaApplyOp.!(instr).source;
+		var point = source.column - 1;
+		var pad = if(source.startColumn != -1, source.startColumn-1, point);
+		extractTmpConstraint(pad);
+		tmpBuf.pad(' ', pad);
+		tmpBuf.pad('~', point);
```

Comment:
I was thinking of centering the value with a format of `|----(val)----|` or `^^^^ val ^^^^`. Wdyt?

Reply:
I would like to use this one: `|----(val)----|`